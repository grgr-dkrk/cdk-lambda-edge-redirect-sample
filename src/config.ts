import * as dotenv from 'dotenv'

dotenv.config()

/**
 * account & region
 */
export const AWS_ACCOUNT = process.env.AWS_ACCOUNT as string
export const AWS_REGION = process.env.AWS_REGION ?? 'us-east-1'
export const env = {
  account: AWS_ACCOUNT,
  region: AWS_REGION,
}

/**
 * Environment
 */
export const ENVIRONMENT_STAGING = 'staging' as const
export const ENVIRONMENT_PRODUCTION = 'production' as const
export type Environment =
  | typeof ENVIRONMENT_STAGING
  | typeof ENVIRONMENT_PRODUCTION

/**
 * Domain
 */
export const DOMAIN_STAGING = process.env.DOMAIN_STAGING as string
export const DOMAIN_PRODUCTION = process.env.DOMAIN_PRODUCTION as string
export type DomainRecord = Record<
  typeof DOMAIN_STAGING | typeof DOMAIN_PRODUCTION,
  string
>
export const domain: DomainRecord = {
  [ENVIRONMENT_STAGING]: DOMAIN_STAGING,
  [ENVIRONMENT_PRODUCTION]: DOMAIN_PRODUCTION,
}

/**
 * HostedZone
 */
export const HOSTED_ZONE_STAGING = process.env.HOSTED_ZONE_STAGING as string
export const HOSTED_ZONE_PRODUCTION = process.env
  .HOSTED_ZONE_PRODUCTION as string
export const hostedZone: Record<Environment, string> = {
  [ENVIRONMENT_STAGING]: HOSTED_ZONE_STAGING,
  [ENVIRONMENT_PRODUCTION]: HOSTED_ZONE_PRODUCTION,
}

/**
 * Policy
 */
export const USER_ARN_STAGING = process.env.USER_ARN_STAGING as string
export const USER_ARN_PRODUCTION = process.env.USER_ARN_PRODUCTION as string
export const principal: Record<Environment, string> = {
  [ENVIRONMENT_STAGING]: USER_ARN_STAGING,
  [ENVIRONMENT_PRODUCTION]: USER_ARN_PRODUCTION,
}

/**
 * Assets
 */
export const ASSETS_WEBSITE = './website' as const
