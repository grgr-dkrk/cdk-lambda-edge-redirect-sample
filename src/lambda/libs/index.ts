import { hasExtension } from './hasExtension'
import { hasTrailingSlash } from './hasTrailingSlash'
import { createRedirectResponse } from './createResponse'

export { createRedirectResponse, hasExtension, hasTrailingSlash }
