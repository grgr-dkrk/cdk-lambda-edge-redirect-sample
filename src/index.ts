#!/usr/bin/env node
import 'source-map-support/register'
import { App } from '@aws-cdk/core'
import { DeploymentStack } from './stacks/deployment-stack'
import { Environment, env, domain, hostedZone } from './config'
import { HostedZoneStack } from './stacks/hostedZone-stack'
import { CertificateManagerStack } from './stacks/certificate-manager-stack'
import { CloudFrontStack } from './stacks/cloudfront-stack'
import { resolve } from 'path'
import { Source } from '@aws-cdk/aws-s3-deployment'

const app = new App()

const environment = app.node.tryGetContext('environment') as
  | Environment
  | undefined

if (!env.account) throw new Error('not found: account')
if (!env.region) throw new Error('not found: region')
if (!environment) throw new Error('not found: environment')

const domainName = domain[environment]
const hostedZoneName = hostedZone[environment]

/**
 * HostedZoneStack
 */
const { zone } = new HostedZoneStack(app, `HostedZoneStack-${environment}`, {
  hostedZoneName,
  env,
})

/**
 * CertificateManagerStack
 */
const { certificate } = new CertificateManagerStack(
  app,
  `CertificateManagerStack-${environment}`,
  {
    hostedZoneName,
    zone,
    env,
  },
)

/**
 * CloudFrontStackStack
 */
const cloudFront = new CloudFrontStack(app, `CloudFrontStack-${environment}`, {
  environment,
  domainName,
  certificate,
  zone,
  env,
})

/**
 * DeploymentStack
 */
new DeploymentStack(app, `DeploymentStack-${environment}`, {
  cloudFront,
  source: Source.asset(`${resolve(__dirname)}/website/`),
  env,
})
