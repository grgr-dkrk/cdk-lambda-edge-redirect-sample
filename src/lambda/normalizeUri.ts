import { hasExtension, hasTrailingSlash } from './libs'

/**
 * normalizeUri
 * @param { string } uri - URI
 * @return { string } newUri - new URI
 */
export const normalizeUri = (uri: string): string => {
  const INDEX_PATH = 'index.html' as const

  // if uri has some extension, return uri as it is.
  if (hasExtension(uri)) return uri

  // return `${uri}/index.html`
  return hasTrailingSlash(uri) ? `${uri}${INDEX_PATH}` : `${uri}/${INDEX_PATH}`
}
