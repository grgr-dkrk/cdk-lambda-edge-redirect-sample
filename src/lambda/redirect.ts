import { CloudFrontResultResponse } from 'aws-lambda'
import { createRedirectResponse } from './libs'

/**
 * redirect
 * @param { string } uri - URI
 * @return { Response | null } response | null - Response for Redirect or Null
 */
export const checkRedirectResponse = (
  uri: string,
): CloudFrontResultResponse | null => {
  // if uri is /foo/*, redirect to /bar/.
  if (uri.match(/^\/foo\/?/)) {
    return createRedirectResponse('/bar/')
  }

  return null
}
