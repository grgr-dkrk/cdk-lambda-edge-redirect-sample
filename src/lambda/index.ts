import { CloudFrontRequestEvent, CloudFrontRequestResult } from 'aws-lambda'
import { normalizeUri } from './normalizeUri'
import { checkRedirectResponse } from './redirect'

export const lambdaFunctions = async (
  event: CloudFrontRequestEvent,
): Promise<CloudFrontRequestResult> => {
  const request = event.Records[0].cf.request
  const { uri } = request

  // redirect
  const redirectResponse = checkRedirectResponse(uri)
  if (redirectResponse) {
    return redirectResponse
  }

  // rewrite
  const normalizedUri = normalizeUri(uri)
  return { ...request, uri: normalizedUri }
}
