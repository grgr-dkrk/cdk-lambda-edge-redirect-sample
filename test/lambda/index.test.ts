import { CloudFrontRequestEvent } from 'aws-lambda'
import { lambdaFunctions } from '../../src/lambda'

const createLambdaFunctionMock = (uri: string): CloudFrontRequestEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: '',
          distributionId: '',
          eventType: 'viewer-request',
          requestId: '',
        },
        request: {
          clientIp: '',
          method: 'GET',
          querystring: '',
          uri,
          headers: {},
        },
      },
    },
  ],
})

describe('lambdaFunctions', () => {
  it('snapshot: /foo', async () => {
    expect(
      await lambdaFunctions(createLambdaFunctionMock('/foo')),
    ).toMatchSnapshot()
  })
  it('snapshot: /foo/', async () => {
    expect(
      await lambdaFunctions(createLambdaFunctionMock('/foo/')),
    ).toMatchSnapshot()
  })
  it('snapshot: /foo/index.html', async () => {
    expect(
      await lambdaFunctions(createLambdaFunctionMock('/foo/index.html')),
    ).toMatchSnapshot()
  })
  it('snapshot: /foo/baz/', async () => {
    expect(
      await lambdaFunctions(createLambdaFunctionMock('/foo/baz/')),
    ).toMatchSnapshot()
  })
  it('snapshot: /foo/?user=dkrk', async () => {
    expect(
      await lambdaFunctions(createLambdaFunctionMock('/foo/?user=dkrk')),
    ).toMatchSnapshot()
  })
  it('snapshot: /foo/index.html?user=dkrk', async () => {
    expect(
      await lambdaFunctions(
        createLambdaFunctionMock('/foo/index.html?user=dkrk'),
      ),
    ).toMatchSnapshot()
  })
})
