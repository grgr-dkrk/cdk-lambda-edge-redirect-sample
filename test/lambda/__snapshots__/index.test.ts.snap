// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambdaFunctions snapshot: /foo 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;

exports[`lambdaFunctions snapshot: /foo/ 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;

exports[`lambdaFunctions snapshot: /foo/?user=dkrk 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;

exports[`lambdaFunctions snapshot: /foo/baz/ 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;

exports[`lambdaFunctions snapshot: /foo/index.html 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;

exports[`lambdaFunctions snapshot: /foo/index.html?user=dkrk 1`] = `
Object {
  "headers": Object {
    "location": Array [
      Object {
        "key": "Location",
        "value": "/bar/",
      },
    ],
  },
  "status": "301",
  "statusDescription": undefined,
}
`;
